rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,655 @@
+  BotInstanceSpec spec = 5;
+  // Fields that are set by the server as results of operations. These should
+  // not be modified by users.
+  BotInstanceStatus status = 6;
+}
+
+message BotInstanceSpec {
+  // Empty as is not user configurable.
+  // Eventually this could be leveraged for simple command and control?
+}
+
+// BotInstanceStatusHeartbeat contains information self-reported by an instance
+// of a Bot. This information is not verified by the server and should not be
+// trusted.
+message BotInstanceStatusHeartbeat {
+  // The timestamp that the heartbeat was recorded by the Auth Server. Any
+  // value submitted by `tbot` for this field will be ignored.
+  google.protobuf.Timestamp recorded_at = 1;
+  // Indicates whether this is the heartbeat submitted by `tbot` on startup.
+  bool is_startup = 2;
+  // The version of `tbot` that submitted this heartbeat.
+  string version = 3;
+  // The hostname of the host that `tbot` is running on.
+  string hostname = 4;
+  // The duration that `tbot` has been running for when it submitted this
+  // heartbeat.
+  google.protobuf.Duration uptime = 5;
+  // The currently configured join_method.
+  string join_method = 6;
+  // Indicates whether `tbot` is running in one-shot mode.
+  bool one_shot = 7;
+  // In future iterations, additional information can be submitted here.
+  // For example, the configuration of `tbot` or the health of individual
+  // outputs.
+}
+
+// BotInstanceStatusAuthentication contains information about a join or renewal.
+// Ths information is entirely sourced by the Auth Server and can be trusted.
+message BotInstanceStatusAuthentication {
+  // The timestamp that the join or renewal was authenticated by the Auth
+  // Server.
+  google.protobuf.Timestamp authenticated_at = 1;
+  // The join method used for this join or renewal.
+  string join_method = 2;
+  // The join token used for this join or renewal. This is only populated for
+  // delegated join methods as the value for `token` join methods is sensitive.
+  string join_token = 3;
+  // The metadata sourced from the join method.
+  google.protobuf.Struct metadata = 4;
+  // On each renewal, this generation is incremented. For delegated join
+  // methods, this counter is not checked during renewal. For the `token` join
+  // method, this counter is checked during renewal and the Bot is locked out if
+  // the counter in the certificate does not match the counter of the last
+  // authentication.
+  int32 generation = 5;
+}
+
+// BotInstanceStatus holds the status of a BotInstance.
+message BotInstanceStatus {
+  // The unique identifier for this bot.
+  string id = 1;
+  // The public key of the Bot instance.
+  // When authenticating a Bot instance, the full public key must be compared
+  // rather than just the fingerprint to mitigate pre-image attacks.
+  bytes public_key = 2;
+  // The fingerprint of the public key of the Bot instance.
+  string fingerprint = 3;
```

Comment:
I've so far kept the public key and fingerprint fields even though we're adding a UUID. Do we want to keep these? I don't think they're strictly required, but may still be useful.

Reply:
I think it's useful info but if it's placed at the top level, we don't have a context point for when the value was this. We could move this into `BotInstanceStatusAuthentication` and that way we'll properly handle the key changing.

Reply:
Good call, will move it into that message.